### ISSUE TITLE
Non-uniform histogram

### ISSUE THREAD

**Issue by arthurprs** [posix4e/rust-metrics#81]
Current histogram is a great uniform histogram but it's a poor metrics histogram most of the time, as the first data point has the same weight as the most recent data point.

This lib should provide some sort of decaying histogram as well.

**Comment by posix4e** [comment]
@arthurprs  stupid question, any reason why we just wouldn't use two histograms, each half the size. One for short term one for long. Can you provide some examples?

**Comment by posix4e** [comment]
@brayniac  might have some opinions here

**Comment by brayniac** [comment]
@arthurprs - I'm curious if you know of a stats library that handles histograms that way. What I've seen is that histograms are latched over some intergation window - eg record for 60s, extract data, clear histogram, record. This gives you access to the percentiles and other derived metrics for that minute. This is how I did it for my time-interval stats library (https://github.com/brayniac/tic)

**Comment by arthurprs** [comment]
My experience is probably biased but most libraries that I know provide or even default to a decaying histogram.

Java: https://github.com/dropwizard/metrics/blob/3.2-development/metrics-core/src/main/java/com/codahale/metrics/MetricRegistry.java#L525

Python: https://github.com/omergertel/pyformance/blob/master/pyformance/meters/histogram.py#L22

C#: https://github.com/Recognos/Metrics.NET/blob/2e7e3ba625298bd8dc35e577122a5285d3a50619/Src/Metrics/MetricsConfig.cs#L29

**Comment by arthurprs** [comment]
If you consider the prometheus style libraries the "decay sampling" Histograms are called Summaries instead, sort of.

**Comment by brayniac** [comment]
@arthurprs - cool! Thanks for the links. I need to do some reading before I have more to say. I'm curious to learn when decaying might be a better fit than latched.

**Comment by sfackler** [comment]
I ended up implementing a copy of Java Metric's `ExponentiallyDecayingReservoir` to swap out the HDR histogram that metrics uses by default - snapshots are about 100x faster in addition to having time decay. I can push it upstream if people are interested.

**Comment by arthurprs** [comment]
Please do, if you put it in GH I can cherry-pick. I'm not sure @posix4e position since he marked the crate deprecated.

**Comment by brayniac** [comment]
I wonder if this is something we should support in tic. Would appreciate benchmarks/links.

**Comment by arthurprs** [comment]
I posted links up thread.

**Comment by brayniac** [comment]
@arthurprs - sorry, links/benches request was for @sfackler - was replying from mobile =)

**Comment by sfackler** [comment]
Here's the histogram implementation: https://gist.github.com/sfackler/2ea5c5b2ee0cc1447b9ae415292adaba

I don't have the exact code I used for the benchmark comparison, but it was IIRC just loading up an HDR Histogram and the histogram in the gist with default settings, filling them with `0..1000` and then generating a snapshot (count, min, max, mean, stddev, p50, p75,  p95, p98, p99, p999) in the `Bencher::iter` closure. The average times were 3,544,088 ns for HDR vs 36,804 ns for the new one.

**Comment by posix4e** [comment]
Neat I'd be willing to take it as a pr

**Comment by sfackler** [comment]
Update times are a bit slower since it has to pull the current time but it's still decently fast - maybe 100ns? Don't have hard numbers for that written down anywhere.

**Comment by sfackler** [comment]
Cool, I'll push one up today.

**Comment by brayniac** [comment]
@sfackler - thanks. I hacked together a quick benchmark for update operation for the decaying histogram proposed: https://github.com/brayniac/histobench

TLDR:
```
     Running target/release/deps/bench_decaying-c16fb48d5628f0a5

running 1 test
test decaying_increment ... bench:         119 ns/iter (+/- 9)

test result: ok. 0 passed; 0 failed; 0 ignored; 1 measured; 0 filtered out

     Running target/release/deps/bench_histogram-984d88d3e7dfe1ad

running 1 test
test brayniac_increment ... bench:           7 ns/iter (+/- 0)

test result: ok. 0 passed; 0 failed; 0 ignored; 1 measured; 0 filtered out
```

Percentiles in my histogram crate should be fairly cheap - and ideally we can handle those kinds of things without blocking the threads we're trying to get metrics from. The cost per increment here might be too high for something like tic - 119ns limits us to no more than 10M increments/s - staying closer to 10ns allows for up to 100M/s. 

EDIT: looks like calculating p90 and stddev is a lot cheaper with the proposed decaying histogram. histobench repo updated with additional coverage

**Comment by sfackler** [comment]
Oh yeah, if you have very update heavy workloads then exponential time decay is not going to work well.

**Comment by sfackler** [comment]
I published this at https://crates.io/crates/exponential-decay-histogram

**Comment by posix4e** [comment]
neat
